File: macos/Tests/VoiceKeyCoreTests/WindowPresentationTests.swift
import Testing
@testable import VoiceKeyCore

/// The main window's date line, engine pill, and permission popover.
@Suite struct WindowPresentationTests {

    // MARK: - Factories

    private let calendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC") ?? .gmt
        return calendar
    }()

    /// Wednesday, 5 August 2026, 15:13 UTC.
    private let now = Date(timeIntervalSince1970: 1_785_942_780)

    private func presentation(status: DictationStatus = .idle,
                              microphone: Grant = .granted,
                              accessibility: Grant = .granted,
                              model: Grant = .granted) -> WindowPresentation {
        WindowPresentation.make(status: status, now: now, calendar: calendar,
                                microphone: microphone, accessibility: accessibility, model: model)
    }

    // MARK: - Header

    @Test func theDateLineNamesTheWeekdayAndTheDay() {
        #expect(presentation().dateLine == "Wednesday, 5 August")
    }

    // MARK: - The engine pill

    @Test func thePillFollowsWhatTheAppIsDoing() {
        #expect(presentation(status: .idle).pill.label == "Ready")
        #expect(presentation(status: .recording(elapsed: 3)).pill.label == "Listening…")
        #expect(presentation(status: .transcribing).pill.label == "Transcribing…")
        #expect(presentation(status: .loading).pill.label == "Preparing…")
    }

    @Test func eachStateHasItsOwnToneSoTheColourCanBeReadWithoutTheWord() {
        #expect(presentation(status: .idle).pill.tone == .ready)
        #expect(presentation(status: .recording(elapsed: 3)).pill.tone == .listening)
        #expect(presentation(status: .transcribing).pill.tone == .working)
        #expect(presentation(status: .loading).pill.tone == .working)
    }

    @Test func anErrorPillCarriesTheMessageItself() {
        let pill = presentation(status: .error("Microphone blocked")).pill

        #expect(pill.label == "Microphone blocked")
        #expect(pill.tone == .blocked)
    }

    @Test func onlyAnErrorPillIsWorthClickingThrough() {
        #expect(presentation(status: .error("Microphone blocked")).pill.isActionable)
        #expect(!presentation(status: .idle).pill.isActionable)
        #expect(!presentation(status: .recording(elapsed: 1)).pill.isActionable)
    }

    // MARK: - Permissions

    @Test func everyPermissionIsListedWithItsState() {
        #expect(presentation().permissions.map(\.text)
                    == ["microphone · granted", "accessibility · granted", "model · loaded"])
    }

    @Test func aMissingGrantSaysSoAndAsksForAttention() {
        let lines = presentation(microphone: .denied, accessibility: .unknown, model: .pending).permissions

        #expect(lines.map(\.text)
                    == ["microphone · denied", "accessibility · not granted", "model · loading"])
        #expect(lines.map(\.needsAttention) == [true, true, true])
    }

    @Test func grantedPermissionsAskForNothing() {
        #expect(presentation().permissions.map(\.needsAttention) == [false, false, false])
    }

    @Test func eachLineKnowsWhichSubjectItSpeaksForSoItCanBeActedOn() {
        #expect(presentation().permissions.map(\.subject) == [.microphone, .accessibility, .model])
    }

    // MARK: - The onboarding strip

    @Test func theOnboardingStripIsThereForSomeoneWhoHasNotDictatedYet() {
        #expect(WindowPresentation.showsOnboarding(dismissed: false, transcripts: 0))
    }

    @Test func dismissingTheStripKeepsItAway() {
        #expect(!WindowPresentation.showsOnboarding(dismissed: true, transcripts: 0))
    }

    @Test func theStripRetiresItselfAfterTheThirdDictation() {
        #expect(WindowPresentation.showsOnboarding(dismissed: false, transcripts: 2))
        #expect(!WindowPresentation.showsOnboarding(dismissed: false, transcripts: 3))
    }
}

File: windows/VoiceKey.Core.Tests/WindowPresentationTests.cs
namespace VoiceKey.Core.Tests;

/// <summary>The main window's date line, engine pill, and permission popover.</summary>
public class WindowPresentationTests
{
    // MARK: - Factories

    private static readonly TimeZoneInfo Zone = TimeZoneInfo.Utc;

    /// <summary>Wednesday, 5 August 2026, 15:13 UTC.</summary>
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_785_942_780);

    private static WindowPresentation Present(DictationStatus? status = null,
                                              Grant microphone = Grant.Granted,
                                              Grant model = Grant.Granted) =>
        WindowPresentation.Make(status ?? DictationStatus.Idle, Now, Zone, microphone, model);

    // MARK: - Header

    [Fact]
    public void TheDateLineNamesTheWeekdayAndTheDay()
    {
        Assert.Equal("Wednesday, 5 August", Present().DateLine);
    }

    // MARK: - The engine pill

    [Fact]
    public void ThePillFollowsWhatTheAppIsDoing()
    {
        Assert.Equal("Ready", Present(DictationStatus.Idle).Pill.Label);
        Assert.Equal("Listening…", Present(DictationStatus.Recording(3)).Pill.Label);
        Assert.Equal("Transcribing…", Present(DictationStatus.Transcribing).Pill.Label);
        Assert.Equal("Preparing…", Present(DictationStatus.Loading).Pill.Label);
    }

    [Fact]
    public void EachStateHasItsOwnToneSoTheColourCanBeReadWithoutTheWord()
    {
        Assert.Equal(StatusTone.Ready, Present(DictationStatus.Idle).Pill.Tone);
        Assert.Equal(StatusTone.Listening, Present(DictationStatus.Recording(3)).Pill.Tone);
        Assert.Equal(StatusTone.Working, Present(DictationStatus.Transcribing).Pill.Tone);
        Assert.Equal(StatusTone.Working, Present(DictationStatus.Loading).Pill.Tone);
    }

    [Fact]
    public void AnErrorPillCarriesTheMessageItself()
    {
        var pill = Present(DictationStatus.Error("Microphone blocked")).Pill;

        Assert.Equal("Microphone blocked", pill.Label);
        Assert.Equal(StatusTone.Blocked, pill.Tone);
    }

    [Fact]
    public void OnlyAnErrorPillIsWorthClickingThrough()
    {
        Assert.True(Present(DictationStatus.Error("Microphone blocked")).Pill.IsActionable);
        Assert.False(Present(DictationStatus.Idle).Pill.IsActionable);
        Assert.False(Present(DictationStatus.Recording(1)).Pill.IsActionable);
    }

    // MARK: - Permissions

    [Fact]
    public void EveryPermissionIsListedWithItsState()
    {
        // Windows has no accessibility grant — SendInput needs no permission — so
        // the list is one line shorter than on macOS.
        Assert.Equal(["microphone · granted", "model · loaded"],
            Present().Permissions.Select(line => line.Text));
    }

    [Fact]
    public void AMissingGrantSaysSoAndAsksForAttention()
    {
        var lines = Present(microphone: Grant.Denied, model: Grant.Pending).Permissions;

        Assert.Equal(["microphone · denied", "model · loading"], lines.Select(line => line.Text));
        Assert.Equal([true, true], lines.Select(line => line.NeedsAttention));
    }

    [Fact]
    public void AnUnestablishedGrantReadsAsMissing()
    {
        var lines = Present(microphone: Grant.Unknown, model: Grant.Unknown).Permissions;

        Assert.Equal(["microphone · denied", "model · unavailable"], lines.Select(line => line.Text));
    }

    [Fact]
    public void GrantedPermissionsAskForNothing()
    {
        Assert.Equal([false, false], Present().Permissions.Select(line => line.NeedsAttention));
    }

    [Fact]
    public void EachLineKnowsWhichSubjectItSpeaksForSoItCanBeActedOn()
    {
        Assert.Equal([Subject.Microphone, Subject.Model],
            Present().Permissions.Select(line => line.Subject));
    }

    // MARK: - The onboarding strip

    [Fact]
    public void TheOnboardingStripIsThereForSomeoneWhoHasNotDictatedYet()
    {
        Assert.True(WindowPresentation.ShowsOnboarding(dismissed: false, transcripts: 0));
    }

    [Fact]
    public void DismissingTheStripKeepsItAway()
    {
        Assert.False(WindowPresentation.ShowsOnboarding(dismissed: true, transcripts: 0));
    }

    [Fact]
    public void TheStripRetiresItselfAfterTheThirdDictation()
    {
        Assert.True(WindowPresentation.ShowsOnboarding(dismissed: false, transcripts: 2));
        Assert.False(WindowPresentation.ShowsOnboarding(dismissed: false, transcripts: 3));
    }
}
